### PR TITLE
[12.0] [IMP] sentry: migrate sentry-raven to new api sentry-sdk

### PR DESCRIPTION
**Pull request by fernandahf**
Description of the issue/feature this PR addresses:

Current behavior before PR: events sent for sentry-raven with context, excluding loggers, exceptions and filtering sense data. In fact, Sentry warns about migrate to new api: 

image

Desired behavior after PR is merged: events sent for sentry-sdk with context, excluding loggers, exceptions and filtering sensitive data, show events in sentry sdk without warning about update api. Sentry just shows sentry-sdk version without warning migrate.

image

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: sentry/__init__.py
Comment (moylop260) [inline sentry/__init__.py]:
```suggestion
    """Setup an instance of :class:`sentry_sdk`. 
    :param config: Sentry configuration"""
```

File: sentry/__init__.py
Comment (moylop260) [inline sentry/__init__.py]: In fact, to initialize these kind of feature we need to be sure that the module is loading in odoo

```suggestion
def post_load():
    _logger.info("Initializing sentry...")
    initialize_sentry(odoo_config)
```

And it should be defined in the hooks of the manifest:

```diff
diff --git a/sentry/__manifest__.py b/sentry/__manifest__.py
index 8ef85da1..732cd858 100644
--- a/sentry/__manifest__.py
+++ b/sentry/__manifest__.py
@@ -22,4 +22,5 @@
     'depends': [
         'base',
     ],
+    'post_load': 'post_load',
 }
```

It will fix the following error loading modules for the longpolling:

```txt
Traceback (most recent call last):                                                                                                                                                                                                                   [136/1808]
  File "/home/odoo/instance/odoo/odoo/modules/registry.py", line 60, in __new__
    return cls.registries[db_name]
  File "/home/odoo/instance/odoo/odoo/tools/func.py", line 69, in wrapper
    return func(self, *args, **kwargs)
  File "/home/odoo/instance/odoo/odoo/tools/lru.py", line 44, in __getitem__
    a = self.d[obj].me
KeyError: 'DATABASENAME'

During handling of the above exception, another exception occurred:
```

You can use another commit for that with `git commit --author="Moisés López <moylop260@vaxuoo.com>"` :)

File: sentry/__manifest__.py
Comment (moylop260) [inline sentry/__manifest__.py:20]: I think we could be added in the `author` section like "Vauxoo"

And you can add the key [`"maintainers": ["your-github-login"],`](https://github.com/OCA/maintainer-tools/blob/f1b559574648c01558b543e7550f5b9be7c92784/template/module/__manifest__.py#L13) to add original authors and we both

**Review by moylop260 — CHANGES_REQUESTED** [review]

**Comment by moylop260** [comment]
Could you check if we can remove `raven` entirely as dependency, please?

**Comment by moylop260** [comment]
@barsi @naglis @Versada
We are migrating the new api of sentry
Could you check it, please?

It is still WIP but if you can get a look here we will grateful

**Comment by moylop260** [comment]
For the record,
A good improvement I saw using `sentry-sdk` instead of `sentry-raven` is it is adding a tag with the hash of the trace
 - Screen Shot 2020-11-30 at 20 18 57

I had needed before to have this value and I just use the following PR to add it manually:
 - https://github.com/Vauxoo/odoo/pull/358

Now it is auto-generated.

**Comment by moylop260** [comment]
For the record,
Another improvement:
 - Before of this PR  `USER: IP Address 127.0.0.1`
 - After this PR `USER: IP Address "REAL-PUBLIC-IP"`

File: sentry/logutils.py
Comment (moylop260) [inline sentry/logutils.py:51]: @fernandahf 
Could you check how to add the login and uid using new package, please?

I mean,
 - Screen Shot 2020-11-30 at 21 23 16

Now, it is not added

Comment (moylop260) [reply sentry/logutils.py:51]: FYI I just found the documentation for this:
 - https://docs.sentry.io/platforms/python/guides/django/enriching-events/identify-user/

File: sentry/__manifest__.py
Comment (fernandahf) [reply sentry/__manifest__.py:20]: Done

File: sentry/logutils.py
Comment (fernandahf) [reply sentry/logutils.py:51]: Done

File: sentry/__init__.py
Comment (fernandahf) [reply sentry/__init__.py]: Done

File: sentry/__manifest__.py
Comment (fernandahf) [reply sentry/__manifest__.py:20]: Done

**Comment by fernandahf** [comment]
> Could you check if we can remove `raven` entirely as dependency, please?

File: sentry/logutils.py
Comment (fernandahf) [reply sentry/logutils.py:51]: Done

File: sentry/__init__.py
Comment (hparfr) [inline sentry/__init__.py]: I don't think it's that easy.

Please have a look in the orignal implementation of raven : https://github.com/getsentry/raven-python/blob/d7d14f61b7fb425bcb15512f659626648c494f98/raven/versioning.py#L11


By the way, I don't think it's the responsibility of this module to take assumptions over how find the commit number. I suggest to remove this feature and take the commit identifier from config

File: sentry/const.py
Comment (hparfr) [inline sentry/const.py]: No.
A wrongly configured instance should appear as "production".

File: sentry/processor.py
Comment (hparfr) [inline sentry/processor.py]: url for AUTHORS and LICENSE  ?
Please describe why this copy paste. Are the (new) filtering options not good enough ?

File: sentry/__init__.py
Comment (hparfr) [inline sentry/__init__.py]: Shouldn't it be False by default ?

**Review by hparfr — CHANGES_REQUESTED** [review]
Thanks for this great and needed work.
Didn't had time to test it yet, some remarks from the code review.

File: sentry/__init__.py
Comment (fernandahf) [reply sentry/__init__.py]: Done

File: sentry/__init__.py
Comment (fernandahf) [reply sentry/__init__.py]: I'm agree with respect to the responsability about how find the commit number. I've changed this code like raven code.

File: sentry/const.py
Comment (fernandahf) [reply sentry/const.py]: I've changed 'production' by 'develop'.

File: sentry/__init__.py
Comment (fernandahf) [reply sentry/__init__.py]: Done

File: sentry/processor.py
Comment (fernandahf) [reply sentry/processor.py]: Sentry-sdk does not have a processor funcionality like raven, I've looked for a solution and have found comments about take processor class of raven and to do a custom implementation:

https://github.com/getsentry/sentry-python/issues/211#issuecomment-633101708
https://forum.sentry.io/t/how-to-customize-event-in-before-send/7457

By the way, I've changed this code.

File: sentry/logutils.py
Comment (moylop260) [inline sentry/logutils.py]: this method exists 2 times

File: sentry/processor.py
Comment (moylop260) [inline sentry/processor.py]: I think this method is not required

File: sentry/__init__.py
Comment (moylop260) [inline sentry/__init__.py]: Move all logic from `__init__.py` file to [hooks.py](https://github.com/OCA/maintainer-tools/blob/master/template/module/hooks.py)

And import hooks like:
 - https://github.com/OCA/maintainer-tools/blob/f1b559574648c01558b543e7550f5b9be7c92784/template/module/__init__.py#L7

File: sentry/processor.py
Comment (fernandahf) [reply sentry/processor.py]: Done

File: sentry/__init__.py
Comment (fernandahf) [reply sentry/__init__.py]: Done

File: sentry/logutils.py
Comment (fernandahf) [reply sentry/logutils.py]: Done

**Comment by moylop260** [comment]
@barsi @naglis @versada Could you check it, please?

**Comment by hoangtrann** [comment]
Any updates on this guys?

**Comment by moylop260** [comment]
@hoangtrann 

We are waiting for approves

I mean, we need almost 2 approves to install and configure this module with this new change and do feedback or approve.

Fell free to test it, please

**Comment by hoangtrann** [comment]
@moylop260 let me check this with one of my client's staging instance 😉

**Comment by ryantran-novobi** [comment]
I have deployed and ran this PR successfully on Odoo instance running on Odoo.sh

ping @moylop260 @hparfr @fernandahf

**Comment by kirtangajjar** [comment]
I tested this PR. It works fine on my local. I'm able to see the errors in sentry, however, I'm unable to see anything in the performance screen. Is it expected?

**Comment by hoangtrann** [comment]
image

Not sure if this is relavant, but how could we enhance the presentation of model object in sentry, like `sale.order(124292,)` or `sale.order(124292, 124293)`

@fernandahf @moylop260

**Comment by fernandahf** [comment]
@kirtangajjar

What is the performance screen? I searched but didn't find it.

Regards.

**Comment by kirtangajjar** [comment]
@fernandahf Check the below screenshot.

Screenshot from 2021-04-20 11-02-41

Here are the link to official docs: 

https://docs.sentry.io/product/performance/
https://docs.sentry.io/product/performance/getting-started/

**Comment by kirtangajjar** [comment]
@fernandahf Did you get time to look at it?

**Review by hparfr — APPROVED** [review]
Thanks

**Comment by themreza** [comment]
What's the status here?

**Comment by fernandahf** [comment]
@themreza

We are waiting for approves

I mean, we need almost 1 approve to install and configure this module with this new change and do feedback or approve.

Fell free to test it, please

**Comment by themreza** [comment]
I tested the code in this branch on Odoo 14 and I can confirm that issues are being correctly sent to Sentry.

image

Although the performance feature is sending data, it's really vague and rather pointless.

image

I intentionally ran some excessively slow requests to see the trace, ideally starting from the HTTP request all the way to the database query or the compute methods responsible, but no details were included:

image

Also, there is no frontend tracking via JavaScript:

image

I'm not sure if it's a bug, but when I added `sentry_traces_sample_rate` to `odoo.conf`, the parsed value was `str` rather than `float`.

image

P.S.: It wasn't working at first which threw me off for a while, but I later figured out it was because I was running Odoo via PyCharm's debug mode and it didn't send anything.

**Comment by themreza** [comment]
I figured out the frontend issue and performance tracking.

With a simple template it's possible to inject Sentry's JavaScript bundle:
```
<?xml version="1.0" encoding="UTF-8"?>
<odoo>
    <template id="assets_backend" name="sentry_javascript_assets" inherit_id="web.assets_backend">
        <xpath expr="." position="inside">
            <script
                    src="https://browser.sentry-cdn.com/6.5.1/bundle.tracing.min.js"
                    integrity="sha384-TkhX8dCdk4BiV9tdHwvmIHWhdvw+LGo0SrMWTTBZ2T/sAUnxQWPuRiXELaaiZT6A"
                    crossorigin="anonymous">
            </script>
            <script>
                Sentry.init({
                  dsn: "YOUR_DSN",

                  // This enables automatic instrumentation (highly recommended), but is not
                  // necessary for purely manual usage
                  integrations: [new Sentry.Integrations.BrowserTracing()],

                  // To set a uniform sample rate
                  tracesSampleRate: 0.2
                });
            </script>
        </xpath>
    </template>
</odoo>
```

image

I believe if we enable `propagate_traces` option, we can link the server-sided traces to frontend ones.

But still, the backend traces are not offering a detailed breakdown of what the request consisted of.

**Comment by ryantran-novobi** [comment]
@moylop260 would you mind taking a look at this one? I believe that all the discussions are resolved and it's waiting for your approval so far

**Comment by moylop260** [comment]
@ryantran-novobi 

Let's do it

**Comment by codeagencybe** [comment]
Is this issue already resolved and ready to use now on v14? 
I just downloaded the latest version and installed and it reports to me I need to upgrade the SDK. 
But this issue seems to have resolved that issue? 
Any feedback please?

**Comment by fernandahf** [comment]
Hi @codeagencybe 

It's not in forward-ported yet, but there is one PR related to v14.0

https://github.com/OCA/server-tools/pull/1976

**Pull request merged**
